test/Datasource.test.js

```diff
@@ -1,116 +1,190 @@
-        nonExistantDataSource = new FileReadWriteDatasource(nonExistantFilename);
+describe('Test all datasouses from ../src/lib/datasouce', () => {
+  before(() => {
+      dataSource = new inMemoryDataSource();
+
+
+      //  todo: review this, and add multiply classes structure
+      // const filename = './fixtures/dataCorrect.json';
+      // dataSource = new FileReadWriteDataSource(filename);
+  
+      // const brokenFilename = './fixtures/dataIncorrect.doc';
+      // brokenDataSource = new FileReadWriteDataSource(brokenFilename);
+  
+      // const nonExistantFilename = './Test/fixtures/dataCorrect.doc'    
+      // nonExistantDataSource = new FileReadWriteDataSource(nonExistantFilename);
+
+      // const emptyFilename = './fixtures/dataEmpty.json';    
+      // emptyDataSource = new FileReadWriteDataSource(emptyFilename);
+    });
 
-        const emptyFilename = './fixtures/dataEmpty.json';    
-        emptyDataSource = new FileReadWriteDatasource(emptyFilename);
+  after(() => {
+    console.log('testing END');
+  });
+  
+  describe('#get', ()=> {
+    it('should return Promise ', ()=> {
+      expect(dataSource.get()).toBeA(Promise);
+    });
+
+    it('should resolve with null when key do not exist', ()=> {
+      return dataSource.get('test').then((res) => {
+        expect(res).toEqual(null);  
       });
-      after(() => {
-        console.log('testing END');
+    });
+
+    it('should resolve with value for existing key', ()=> {
+      return dataSource.get('city').then((res) => {
+        expect(res).toEqual('Kyiv');  
       });
-    
-    describe('#get', ()=> {
-        it('should return Promise ', ()=> {
-          expect(dataSource.get()).toBeA(Promise);
-        });
-    
-        it('should resolve with null when key do not exist', ()=> {
-          return dataSource.get('test').then((res) => {
-            expect(res).toEqual(null);  
-          });
+    });
+  })
+  
+  describe('#set', ()=> {
+    it('should return Promise ', ()=> {
+      expect(dataSource.set()).toBeA(Promise);
+    });
+
+    it('should resolve with seted key', ()=> {
+      return dataSource.set('key', 'value').then((res) => {
+        expect(res).toEqual('key');  
+      });
+    });
+  })
+  
+  describe('#delete', ()=> {
+    it('should return Promise ', ()=> {
+      expect(dataSource.delete()).toBeA(Promise);
+    });
+
+    it('should resolve with true', ()=> {
+      return dataSource.delete('key').then((res) => {
+        expect(res).toEqual(true);  
+      });
+    });
+  })
+
+  describe('chaining', () => {
+    it('should return Promise and resolve with a null', () => {
+      return emptyDataSource.get('key').then((res) => {
+        expect(emptyDataSource.get('key')).toBeA(Promise);
+        expect(res).toEqual(null);  
+      });
+    });
+
+    it('set>get shoud return with seted value', () => {
+      return emptyDataSource.set('key', 'value')
+        .then(() =>  emptyDataSource.get('key'))
```

Comment:
at this line `emptyDatasource` seems to be not empty, right? =)
This is the same like following code:
```
let one = 1;
let two = 2;
one++;
two = undefined;
console.log(one + two); // hmm... it should be three, right?
```

Reply:
Soo, its funny))
In this case, I just don't touch emptyDatasource?